util/a2e/atoe.c

```diff
@@ -1342,12 +1343,20 @@ int
 atoe_sprintf(char *buf, char *ascii_chars, ...)
 {
 	int len;
-	char wrkbuf[BUFLEN];
+	int argSize;
 
-	va_list args;
+	va_list args, args_copy;
 	va_start(args, ascii_chars);
+	va_copy(args_copy, args);
+
+	char *ebcdic_chars = a2e_string(ascii_chars);
+	/* Add 1 to accommodate for null terminating char */
+	argSize = vsnprintf(NULL, 0, ebcdic_chars, args_copy) + 1;
+
+	char wrkbuf[argSize];
+	va_end(args_copy);
 
-	len = atoe_vsnprintf(wrkbuf, BUFLEN, ascii_chars, args);
+	len = atoe_vsnprintf(wrkbuf, argSize, ascii_chars, args);
 
 	va_end(args);
 	if (-1 == len) {
```

Comment:
> strcpy((char *)buf, wrkbuf);

Will the `strcpy` segfault if `buf == NULL` (user-input)?

Reply:
> > However, even with your suggested implementation for reducing performance costs, we will at least have to grab the length of the incoming arguments to decide if the "fast-path" option would fail
> 
> Not necessarily. See https://linux.die.net/man/3/vsnprintf.
> 
> > The functions snprintf() and vsnprintf() do not write more than size bytes (including the terminating null byte ('\0')). If the output was truncated due to this limit then the return value is the number of characters (excluding the terminating null byte) which would have been written to the final string if enough space had been available. Thus, **a return value of size or more means that the output was truncated.** (See also below under NOTES.)
> 
> Please confirm if the above statement is also true for our implementation.
> 
> If so, then the following should work:
> 
> ```
> len = atoe_vsnprintf(wrkbuf, BUFLEN, ascii_chars, args);
> if (len >= BUFLEN) {
>    /* Truncated: take the slow path. */
> }
> ```

Size of `wrkbuf` must be defined before the call to `vsnprintf`, this is why we grab the length of the incoming string before creating the char array. If the conditional you suggest passes, we would create another char array with the correct size. Is this acceptable?

Reply:
See https://github.com/eclipse/omr/pull/6601#issuecomment-1185631272.

Reply:
Not sure if it helps or not but with my changes to atoe_vsnprintf in this[1] PR, it would behave more like std vsnprintf! for instance if you call it like `atoe_vsnprintf(NULL, 0, format, arguments)` it would return estimate of how much space that operation would take without actually write to any buffer! (it was returning -1 before my PR)

[1]: https://github.com/eclipse/omr/pull/7158